File: experimental/workflow_mt/producer/rust/src/producer/responses/structc.rs
use super::{hub, identification, producer::Control, protocol, Context};
use crate::stat::Alias;
use crate::test::samples;
use clibri::server;
use uuid::Uuid;

pub enum Response {
    CaseB(protocol::StructB),
    CaseF(protocol::StructF),
    CaseD(protocol::StructD),
}

#[allow(unused_variables)]
pub async fn response<'c, E: server::Error, C: server::Control<E> + Send + Clone>(
    identification: &identification::Identification,
    filter: hub::filter::Filter,
    context: &Context,
    request: &protocol::StructC,
    control: &Control<E, C>,
) -> Result<Response, protocol::StructE> {
    let index = context
        .get_index(identification.uuid(), Alias::StructC)
        .await;
    if index == 1 {
        context
            .inc_stat(identification.uuid(), Alias::StructB)
            .await;
        Ok(Response::CaseB(samples::struct_b::get()))
    } else if index == 2 {
        context
            .inc_stat(identification.uuid(), Alias::StructF)
            .await;
        Ok(Response::CaseF(samples::struct_f::get()))
    } else if index == 3 {
        context
            .inc_stat(identification.uuid(), Alias::StructD)
            .await;
        Ok(Response::CaseD(samples::struct_d::get()))
    } else {
        context
            .inc_stat(identification.uuid(), Alias::StructE)
            .await;
        Err(samples::struct_e::get())
    }
}

File: tests/workflow/producer/typescript (copy)/src/producer/implementation/responses/structc.ts
import {
    Producer,
    Identification,
    Filter,
    broadcastAll,
    Context,
    Protocol,
} from "./index";
import { response } from "../../responses/structc";

export class Response {    
    private _response!: Protocol.StructB | Protocol.StructF | Protocol.StructD | Protocol.StructE;
    private _broadcasts: Array<[string[], Protocol.Convertor<any>]> = [];

    constructor(
        res: Protocol.StructB | Protocol.StructF | Protocol.StructD | Protocol.StructE
    ) {
        this._response = res;
    }

    public broadcast(uuids: string[]): {        
    } {
        const self = this;
        return {            
        };
    }

    public error(): Error | undefined {
        let error: Error | undefined;        
        return error;
    }

    public pack(sequence: number, uuid: string): ArrayBufferLike {
        return this._response.pack(sequence, uuid);
    }

    public broadcasts(): Array<[string[], Protocol.Convertor<any>]> {
        return this._broadcasts;
    }
}

export function handler(
    request: Protocol.StructC,
    consumer: Identification,
    filter: Filter,
    context: Context,
    producer: Producer,
    sequence: number
): Promise<void> {
    return response(request, consumer, filter, context, producer).then(
        (res) => {
            const error: Error | undefined = res.error();
            if (error instanceof Error) {
                return Promise.reject(error);
            }
            return producer
                .send(consumer.uuid(), res.pack(sequence, consumer.uuid()))
                .then(() => {
                    return broadcastAll(producer, res.broadcasts());
                });
        }
    );
}